pingcap/tiflash
# PageStorage skip non continuous sequence safely

### What problem does this PR solve?

Issue Number: close #2317, related to #2312, #2356

Problem Summary:
After #2312 and #2356, we introduce "Formal" PageFiles into `LegacyCompactor::collectPageFilesToCompact`.
So we may read files that are writing by other threads during `LegacyCompactor::collectPageFilesToCompact`. We need further protection on those writable files while applying GC, or we may scan some data in the middle of writing.

### What is changed and how it works?

* Save the meta offset of writable files after `restore` and `write`
* Save the last persisted sequence of writable files after `write`
* Set the last persisted sequence of writable of the overall persisted sequence after `restore`
* When reading meta from those writable files, only read to the meta offset or to the min persisted sequence at maximum
* Add some failpoint to generate non-continuous sequences

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch:

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)
  * Enable `random_slow_page_storage_write` and `random_exception_after_page_storage_sequence_acquired` to generate non-continuous sequence
  * Wait until the meta get compacted by `LegacyCompactor`
  * Restart TiFlash to check whether we can read / write

Side effects



### Release note 

- No release note
